docs_src/design/adr/system-management/001-docker-image-guidelines.md

```diff
@@ -0,0 +1,110 @@
+
+```docker
+security_opt:
+      - "no-new-privileges:true"
+```
+
+- To further prevent privilege escalation attacks the user should be set for the docker container using the `--user=<userid>` or `-u=<userid>` option in their Docker compose file (example below). More details about this flag can be found [here](https://docs.docker.com/engine/reference/run/#user). This follows Rule #4 for Docker security found [here](https://cheatsheetseries.owasp.org/cheatsheets/Docker_Security_Cheat_Sheet.html#rule-2-set-a-user).
+
+```docker
+services:
+  device-virtual:
+    image: ${REPOSITORY}/docker-device-virtual-go${ARCH}:${DEVICE_VIRTUAL_VERSION}
+    user: 4000 # user option using an unprivileged user
+    ports:
+    - "127.0.0.1:49990:49990"
+    container_name: edgex-device-virtual
+    hostname: edgex-device-virtual
+    networks:
+      - edgex-network
+    env_file:
+      - common.env
+    environment:
+      SERVICE_HOST: edgex-device-virtual
+    depends_on:
+      - consul
+      - data
+      - metadata
+```
+
+> NOTE: exception
+    Sometimes containers will require root access to perform their fuctions. For example the System Management Agent requires root access to control other Docker containers. Inthis case you would need to assign the `user: root`.
+
+```docker
+  system:
+    image: edgexfoundry/docker-sys-mgmt-agent-go:1.0.1
+    user: root # user option using root
+    ports:
+      - "48090:48090"
+    container_name: edgex-sys-mgmt-agent
+    hostname: edgex-sys-mgmt-agent
+    networks:
+      edgex-network:
+        aliases:
+            - edgex-sys-mgmt-agent
+    volumes:
+      - db-data:/data/db
+      - log-data:/edgex/logs
+      - consul-config:/consul/config
+      - consul-data:/consul/data
+      - /var/run/docker.sock:/var/run/docker.sock
+    depends_on:
+      - logging
+```
+
+- To avoid a faulty or compromised containers from consuming excess amounts of the host of its resources `resource limits` should be set for each container. more details about `resource limits` can be found [here](https://docs.docker.com/config/containers/resource_constraints/). This follows Rule #4 for Docker security found [here](https://cheatsheetseries.owasp.org/cheatsheets/Docker_Security_Cheat_Sheet.html#rule-7-limit-resources-memory-cpu-file-descriptors-processes-restarts).
+
+- To avoid attackers from writing data to the containers and modifying their files the `--read_only` flag should be set. More details about this flag can be found [here](https://docs.docker.com/compose/compose-file/#domainname-hostname-ipc-mac_address-privileged-read_only-shm_size-stdin_open-tty-user-working_dir). This follows Rule #4 for Docker security found [here](https://cheatsheetseries.owasp.org/cheatsheets/Docker_Security_Cheat_Sheet.html#rule-8-set-filesystem-and-volumes-to-read-only).
```

Comment:
```suggestion
- To avoid attackers from writing data to the containers and modifying their files the `--read_only` flag should be set. More details about this flag can be found [here](https://docs.docker.com/compose/compose-file/#domainname-hostname-ipc-mac_address-privileged-read_only-shm_size-stdin_open-tty-user-working_dir). This follows Rule #8 for Docker security found [here](https://cheatsheetseries.owasp.org/cheatsheets/Docker_Security_Cheat_Sheet.html#rule-8-set-filesystem-and-volumes-to-read-only).
```

Reply:
Will this read-only apply to container's volume or it is just the container's file system? maybe it is worth to mention it.

Reply:
I would say if you consider "volume" something that you attach to a container that holds your persistent data or tmpfs, this option makes the file system generated from the container image read-only -- i.e. read-only unless a writable volume is mounted on top.